### Download and load Data

In [2]:
import pandas as pd
import numpy as np

In [4]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So, as the table above indicates that the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.

Let's see how many data points we have and whether this dataset contains any missing values

In [5]:
concrete_data.shape

(1030, 9)

In [6]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [7]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks clean and is ready to be used for model creation 

#### Split data into predictors and target

In [9]:
concrete_data_columns = concrete_data.columns 

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] 
target = concrete_data['Strength']
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [10]:
predictors.shape[1]

8

In [11]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

### Part A: Create the Baseline Model

In [12]:
# Split the data into train dataset and test dataset 

from sklearn.model_selection import train_test_split

predictors_train, predictors_test, target_train, target_test = train_test_split(predictors, target, test_size=0.3, random_state=4)
print('Train set:', predictors_train.shape, target_train.shape)
print('Test set:', predictors_test.shape, target_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


In [13]:
import keras
from keras.models import Sequential 
from keras.layers import Dense

#### Create a function that defines our regression model for us so that we can conveninetly call it to create our model

In [14]:
# define the model
def regression_model():
    # create model
    model = Sequential()
    n_cols = predictors.shape[1]
    model.add(Dense(10, activation = 'relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above baseline model that has one hidden layer of 10 nodes, a ReLu activation function, Use Adam optimizer and the Mean Squared Error as the loss function.

#### Train and Evaluate the model

In [15]:
# Build the model
model = regression_model()

In [16]:
# fit the model
model.fit(predictors_train, target_train, epochs=50, verbose=2)

Epoch 1/50
 - 0s - loss: 84767.4399
Epoch 2/50
 - 0s - loss: 26470.1070
Epoch 3/50
 - 0s - loss: 5099.2483
Epoch 4/50
 - 0s - loss: 927.7107
Epoch 5/50
 - 0s - loss: 627.3328
Epoch 6/50
 - 0s - loss: 617.1181
Epoch 7/50
 - 0s - loss: 600.3272
Epoch 8/50
 - 0s - loss: 585.2118
Epoch 9/50
 - 0s - loss: 570.5773
Epoch 10/50
 - 0s - loss: 554.5779
Epoch 11/50
 - 0s - loss: 540.5643
Epoch 12/50
 - 0s - loss: 524.4924
Epoch 13/50
 - 0s - loss: 509.9996
Epoch 14/50
 - 0s - loss: 495.7943
Epoch 15/50
 - 0s - loss: 481.0937
Epoch 16/50
 - 0s - loss: 467.4966
Epoch 17/50
 - 0s - loss: 455.2483
Epoch 18/50
 - 0s - loss: 443.0321
Epoch 19/50
 - 0s - loss: 428.6238
Epoch 20/50
 - 0s - loss: 417.6832
Epoch 21/50
 - 0s - loss: 406.4416
Epoch 22/50
 - 0s - loss: 395.9177
Epoch 23/50
 - 0s - loss: 385.6465
Epoch 24/50
 - 0s - loss: 376.3878
Epoch 25/50
 - 0s - loss: 369.2497
Epoch 26/50
 - 0s - loss: 359.6609
Epoch 27/50
 - 0s - loss: 352.4573
Epoch 28/50
 - 0s - loss: 344.0099
Epoch 29/50
 - 0s - loss

In [17]:
# Evaluate the model
pred_loss = model.evaluate(predictors_test, target_test)
pred_loss

309/309 [==============================] - 0s 42us/step


293.05788671005894

#### Repeat step 1 to step 3 for 50 times and create a list of 50 mean squared errors

In [18]:
losses = []
for i in range(50):
        model.fit(predictors_train, target_train, epochs=50,verbose=0)
        pred_loss = model.evaluate(predictors_test, target_test, verbose=0)
        losses.append(pred_loss)
        preds = model.predict(predictors_test)

#### Report the mean and the standard deviation of the mean squared errors

In [20]:
import statistics 
from statistics import stdev 


print("Mdoel A: Mean of losses is % s " % (statistics.mean(losses))) 
print('\n')
print("Model A: Standard Deviation of losses is % s " % (statistics.stdev(losses))) 

Mdoel A: Mean of losses is 88.7262387544206 


Model A: Standard Deviation of losses is 36.13764574668182 


### Part B: Feed the Model Normalized Data 

#### Normalize the data

In [26]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

#### Split the data for training and testing

In [32]:
from sklearn.model_selection import train_test_split
predictors_train, predictors_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=4)
print('Train set:', predictors_train.shape, target_train.shape)
print('Test set:', predictors_test.shape, target_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


#### Train and Evaluate the model

In [23]:
# Build the model
model = regression_model()

In [27]:
# fit the model
model.fit(predictors_train, target_train, epochs=50, verbose=2)

Epoch 1/50
 - 0s - loss: 273.2515
Epoch 2/50
 - 0s - loss: 262.9567
Epoch 3/50
 - 0s - loss: 253.7825
Epoch 4/50
 - 0s - loss: 245.2358
Epoch 5/50
 - 0s - loss: 237.5335
Epoch 6/50
 - 0s - loss: 230.6688
Epoch 7/50
 - 0s - loss: 224.2369
Epoch 8/50
 - 0s - loss: 218.5861
Epoch 9/50
 - 0s - loss: 213.3020
Epoch 10/50
 - 0s - loss: 208.5308
Epoch 11/50
 - 0s - loss: 204.2182
Epoch 12/50
 - 0s - loss: 200.3252
Epoch 13/50
 - 0s - loss: 196.7164
Epoch 14/50
 - 0s - loss: 193.4094
Epoch 15/50
 - 0s - loss: 190.2924
Epoch 16/50
 - 0s - loss: 187.4820
Epoch 17/50
 - 0s - loss: 184.8369
Epoch 18/50
 - 0s - loss: 182.4589
Epoch 19/50
 - 0s - loss: 180.0968
Epoch 20/50
 - 0s - loss: 178.0012
Epoch 21/50
 - 0s - loss: 176.0404
Epoch 22/50
 - 0s - loss: 174.1321
Epoch 23/50
 - 0s - loss: 172.3849
Epoch 24/50
 - 0s - loss: 170.8072
Epoch 25/50
 - 0s - loss: 169.2868
Epoch 26/50
 - 0s - loss: 167.7179
Epoch 27/50
 - 0s - loss: 166.3657
Epoch 28/50
 - 0s - loss: 164.9391
Epoch 29/50
 - 0s - loss: 163

In [28]:
# Evaluate the model
pred_loss = model.evaluate(predictors_test, target_test)
pred_loss

309/309 [==============================] - 0s 45us/step


165.56653933231885

#### Repeat step 1 to step 3 for 50 times and create a list of 50 mean squared errors

In [38]:
losses = []
for i in range(50):
        model.fit(predictors_train, target_train, epochs=50,verbose=0)
        pred_loss = model.evaluate(predictors_test, target_test, verbose=0)
        losses.append(pred_loss)
        preds = model.predict(predictors_test)

#### Report the mean and the standard deviation of the mean squared errors

In [39]:
print("Mdoel B: Mean of losses is % s " % (statistics.mean(losses))) 
print('\n')
print("Model B: Standard Deviation of losses is % s " % (statistics.stdev(losses))) 

Mdoel B: Mean of losses is 45.12774745509077 


Model B: Standard Deviation of losses is 8.38101846035228 


### Part C: Repeat Part B but use 100 epochs for training

#### Train the model using 100 epochs and Evaluate the model

In [44]:
# Build the model
model = regression_model()

In [45]:
# fit the model
model.fit(predictors_train, target_train, epochs=100, verbose=2)

Epoch 1/100
 - 0s - loss: 1586.4364
Epoch 2/100
 - 0s - loss: 1570.3082
Epoch 3/100
 - 0s - loss: 1553.6469
Epoch 4/100
 - 0s - loss: 1536.7785
Epoch 5/100
 - 0s - loss: 1519.1480
Epoch 6/100
 - 0s - loss: 1500.4341
Epoch 7/100
 - 0s - loss: 1481.0301
Epoch 8/100
 - 0s - loss: 1460.2264
Epoch 9/100
 - 0s - loss: 1438.9418
Epoch 10/100
 - 0s - loss: 1415.8290
Epoch 11/100
 - 0s - loss: 1391.9500
Epoch 12/100
 - 0s - loss: 1367.3978
Epoch 13/100
 - 0s - loss: 1340.8440
Epoch 14/100
 - 0s - loss: 1314.3390
Epoch 15/100
 - 0s - loss: 1285.9445
Epoch 16/100
 - 0s - loss: 1257.8093
Epoch 17/100
 - 0s - loss: 1228.1021
Epoch 18/100
 - 0s - loss: 1198.3882
Epoch 19/100
 - 0s - loss: 1168.2329
Epoch 20/100
 - 0s - loss: 1137.5710
Epoch 21/100
 - 0s - loss: 1106.9346
Epoch 22/100
 - 0s - loss: 1075.5451
Epoch 23/100
 - 0s - loss: 1044.4896
Epoch 24/100
 - 0s - loss: 1012.9145
Epoch 25/100
 - 0s - loss: 981.5769
Epoch 26/100
 - 0s - loss: 950.9313
Epoch 27/100
 - 0s - loss: 920.4617
Epoch 28/100


In [46]:
# Evaluate the model
pred_loss = model.evaluate(predictors_test, target_test)
pred_loss

309/309 [==============================] - 0s 81us/step


180.36613084120273

#### Repeat step 1 to step 3 for 50 times and create a list of 50 mean squared errors

In [47]:
losses = []
for i in range(50):
        model.fit(predictors_train, target_train, epochs=50,verbose=0)
        pred_loss = model.evaluate(predictors_test, target_test, verbose=0)
        losses.append(pred_loss)
        preds = model.predict(predictors_test)

#### Report the mean and the standard deviation of the mean squared errors

In [41]:
print("Mdoel C: Mean of losses is % s " % (statistics.mean(losses))) 
print('\n')
print("Model C: Standard Deviation of losses is % s " % (statistics.stdev(losses))) 

Mdoel C: Mean of losses is 41.45293532954836 


Model C: Standard Deviation of losses is 0.35972173483323516 


### Part D: Repeat part B but Build a neural network with 3 hidden layers each of which has 10 nodes and ReLu activation 

In [48]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    n_cols = predictors.shape[1]
    model.add(Dense(10, activation = 'relu', input_shape=(n_cols,)))
    model.add(Dense(10,activation = 'relu'))
    model.add(Dense(10,activation = 'relu'))
    model.add(Dense(1))
    
    # compile the model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [49]:
# Build the model
model = regression_model()

In [51]:
# fit the model
model.fit(predictors_train, target_train, epochs=50, verbose=2)

Epoch 1/50
 - 0s - loss: 103.3315
Epoch 2/50
 - 0s - loss: 101.5237
Epoch 3/50
 - 0s - loss: 99.8790
Epoch 4/50
 - 0s - loss: 98.5594
Epoch 5/50
 - 0s - loss: 96.8684
Epoch 6/50
 - 0s - loss: 95.2787
Epoch 7/50
 - 0s - loss: 93.7645
Epoch 8/50
 - 0s - loss: 91.8484
Epoch 9/50
 - 0s - loss: 90.2057
Epoch 10/50
 - 0s - loss: 88.6616
Epoch 11/50
 - 0s - loss: 87.2427
Epoch 12/50
 - 0s - loss: 85.8461
Epoch 13/50
 - 0s - loss: 84.1557
Epoch 14/50
 - 0s - loss: 83.0485
Epoch 15/50
 - 0s - loss: 81.5035
Epoch 16/50
 - 0s - loss: 80.7381
Epoch 17/50
 - 0s - loss: 79.2688
Epoch 18/50
 - 0s - loss: 78.1710
Epoch 19/50
 - 0s - loss: 77.1134
Epoch 20/50
 - 0s - loss: 76.1611
Epoch 21/50
 - 0s - loss: 75.4553
Epoch 22/50
 - 0s - loss: 74.1809
Epoch 23/50
 - 0s - loss: 72.9289
Epoch 24/50
 - 0s - loss: 71.7492
Epoch 25/50
 - 0s - loss: 70.9521
Epoch 26/50
 - 0s - loss: 69.8057
Epoch 27/50
 - 0s - loss: 68.8191
Epoch 28/50
 - 0s - loss: 67.4558
Epoch 29/50
 - 0s - loss: 66.2634
Epoch 30/50
 - 0s - l

In [52]:
# Evaluate the model
pred_loss = model.evaluate(predictors_test, target_test)
pred_loss

309/309 [==============================] - 0s 52us/step


63.2260259239419

#### Repeat step 1 to step 3 for 50 times and create a list of 50 mean squared errors

In [53]:
losses = []
for i in range(50):
        model.fit(predictors_train, target_train, epochs=50,verbose=0)
        pred_loss = model.evaluate(predictors_test, target_test, verbose=0)
        losses.append(pred_loss)
        preds = model.predict(predictors_test)

#### Report the mean and the standard deviation of the mean squared errors

In [54]:
print("Mdoel D: Mean of losses is % s " % (statistics.mean(losses))) 
print('\n')
print("Model D: Standard Deviation of losses is % s " % (statistics.stdev(losses))) 

Mdoel D: Mean of losses is 35.35724790924961 


Model D: Standard Deviation of losses is 4.14722717846581 


### Conclution of Model Building Experiements with different parameters 

#### Part A:
   - Inputs of Model A are not normalized, neural network has 1 layer can be viewed as a very basic shellow neural network.
   - Mean of MSEs is very high compared to the other models, along wiht high variance in terms of high Standard Deviation of MSEs.

#### Part B:
- Input Data Normalization is helpful to imporve model performance because it can prevent the model giving greater weights on variables with larger values 
- The mean of MSEs is more stable, because Data Normalization brings in a lower standard deviation of MSEs compared to what's seen from part A.
- The mean of MSEs is the same, because the model architecture is the same. 

#### Part C:
- Epoch increased from 50 to 100, and this result in both mean and standard deviation of MSEs reduced significantly at the same time.
- Greater number of epochs means the more data feed in and the longer training time can allow the model learn more patterns  from the inputs.

#### Part D:
    - Mean and Standard Deviation of MSEs are the reduced significantly. 
    - Maybe because the more hidden layers helps.